File: functional/Queens.java
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Queens {
    static int size;

    public static void main(String[] args) {
        if (args.length > 0) size = Integer.valueOf(args[0]);
        System.out.println(queen(size).map(a->a.mapToObj(String::valueOf).collect(Collectors.joining(" "))).collect(Collectors.joining("\n")));

    }

    public static Stream<IntStream> queen(int n) {
        if (n <= 0) {
            return Stream.of(IntStream.empty());
        } else {
            return queen(n - 1).flatMap(Queens::add);
        }
    }

    public static Stream<IntStream> add(IntStream x) {
        int[] xs = x.toArray();
        return IntStream.rangeClosed(1, size).filter(y->
                !Arrays.stream(xs).anyMatch(e->e==y) &&
                !IntStream.range(0, xs.length).map(n->(xs.length-n)-Math.abs(y-xs[n])).anyMatch(e->e==0)
                ).mapToObj(y->IntStream.concat(Arrays.stream(xs), IntStream.of(y)));
    }
}

File: functional/queens.scala
object Queens {
    def main(args: Array[String]) {
        val size = args(0).toInt
        var queen: (Int) => List[List[Int]] = null
        queen = (n) => {
            if (n <= 0) {
                List(List())
            } else {
                queen(n-1).
                flatMap(x=>(1 to size).
                    collect{case y if !x.contains(y) &&
                                      !x.zipWithIndex.map{z=>x.size-z._2-(y-z._1).abs}.contains(0)
                            =>x:+y})
            }
        }
        println(queen(size).map(_.mkString(" ")).mkString("\n"))
    }
}
